### PR TITLE
Add Type Definitions for servicestack-client

### PR DESCRIPTION
**Pull request by mythz**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [ ] Run `npm run lint -- package-name` if a `tslint.json` is present.

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.

### CODE REVIEW

**Comment by mythz** [comment]
bump?

File: servicestack-client/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+export const toDate: (s: string) => Date;
+export const toDateFmt: (s: string) => string;
+export const padInt: (n: number) => string | number;
+export const dateFmt: (d?: Date) => string;
+export const dateFmtHM: (d?: Date) => string;
+export const timeFmt12: (d?: Date) => string;
+}
+
+declare module "servicestack-client" {
+    export = ServiceStack;
```
Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: If `ServiceStack` is also available as a global, can you move this to the top level and instead use an `export as namespace ServiceStack`?

File: servicestack-client/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
+// Type definitions for servicestack-client v0.0.17
+// Project: https://servicestack.net/
+// Definitions by: Demis Bellot <https://github.com/mythz/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference path="../isomorphic-fetch/index.d.ts" />
```
Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: If you move the `export =` to the top level, change this to an `import`

File: servicestack-client/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    eventSourceUrl: string;
+    updateSubscriberUrl: string;
+    eventSourceStop: boolean;
+    constructor(baseUrl: string, channels: string[], options?: any, eventSource?: IEventSourceStatic);
+    onMessage(e: IOnMessageEvent): void;
+    reconnectServerEvents(opt?: any): IEventSourceStatic;
+    invokeReceiver(r: any, cmd: string, el: Element, msg: string, e: any, name: string): void;
+    updateChannels(channels: string[]): void;
+}
+export class HttpMethods {
```
Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: Is this really a class? Do you have a link to the docs for this?

File: servicestack-client/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+export class ErrorResponse {
+    responseStatus: ResponseStatus;
+}
+export interface ISseCommand {
+    userId: string;
+    displayName: string;
+    channels: string;
+    profileUrl: string;
+}
+export interface ISseHeartbeat extends ISseCommand {
```
Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: Each of these interfaces (`ISseCommand`, `ISseHeartbeat`, `ISseJoin`, etc.) are structurally assignable to each other, just as a heads up. Not sure if it matters all that much though.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: servicestack-client/index.d.ts (hunk 3 above)
Comment (mythz) [inline servicestack-client/index.d.ts]: Not sure what you mean? This .d.ts class was generated by tsc from the [servicestack-client index.ts file](https://github.com/ServiceStack/servicestack-client/blob/master/src/index.ts#L235)

File: servicestack-client/index.d.ts (hunk 4 above)
Comment (mythz) [inline servicestack-client/index.d.ts]: It also serves to identify the type of message.

File: servicestack-client/index.d.ts (hunk 1 above)
Comment (mythz) [inline servicestack-client/index.d.ts]: I only need it available from the **servicestack-client** module, e.g:

```ts
import { JsonServiceClient } from 'servicesack-client';
```

Don't really want it under the `ServiceStack` namespace as well.

Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: In that case you shouldn't need to wrap the module in a `declare namespace`, nor should you need to have an `export =` - just export everything from the top level of the module (i.e. keep the `.d.ts` output the same as when TypeScript generates it).

Comment (DanielRosenwasser) [inline servicestack-client/index.d.ts]: In that case you shouldn't need to wrap the module in a `declare namespace`, nor should you need to have an `export =` - just export everything from the top level of the module (i.e. keep the `.d.ts` output the same as when TypeScript generates it).

File: servicestack-client/index.d.ts (hunk 3 above)
Comment (mythz) [inline servicestack-client/index.d.ts]: Because packaging types with npm package [doesn't work with JSPM yet](https://github.com/Microsoft/TypeScript/issues/6012) from what I understand publishing `@types` separately was the only solution. I've just learnt of another approach which I still have to try out, but otherwise I'd need to publish the type defs to `@types`.

**Comment by mythz** [comment]
Looks like we can install `servicestack-client` with jspm and can then re-import the same npm package as a `devDependencies` and tsc was able to find the type defs installed with the project without conflicts so it doesn't look like we'd need to re-publish the type defs here as well.

**Comment by DanielRosenwasser** [comment]
Also note that path mappings will work for your JSPM users:

```json
{
    "compilerOptions": {
        // ...
    },
    "baseUrl": "./jspm_packages",
    "paths": {
        "servicestack-client/*": ["npm/servicestack-client@0.0.17/*"]
    }
}
```

**Comment by DanielRosenwasser** [comment]
Also, sorry it took a few days to check out your PR @mythz - keep in mind that the last few days've been the weekend as well as a national holiday, so not everyone has been available to check out your PR.

**Comment by DanielRosenwasser** [comment]
Ah, [I see from Twitter](https://twitter.com/eamodio/status/803420598428954624) that JSPM apparently removes non-JS files, so your solution may be better.

**Comment by mythz** [comment]
@DanielRosenwasser thanks for the info although it doesn't look like jspm removes non-js files as after installing `servicestack-client` I end up with the directory structure:

```
jspm_packages/
    npm/
        servicestack-client@0.0.17/
            src/
                index.d.ts
                index.js
                index.js.map
                index.ts
```

But adding the above config, e.g:

```json
    "baseUrl": "./jspm_packages",
    "paths": {
        "servicestack-client/*": ["npm/servicestack-client@0.0.17/src/*"]
    }
```

Didn't work initially so I needed to remove `jspm_packages` from the `excludes` list which then had errors trying to build an unrelated package:

```
jspm_packages/npm/iconv-lite@0.4.15/lib/index.d.ts(6,25): error TS2307: Cannot find module 'stream'.
jspm_packages/npm/iconv-lite@0.4.15/lib/index.d.ts(14,32): error TS2304: Cannot find name 'Buffer'.
...
```

I don't intend to base my solution requiring maintaining a manual black list but I added `jspm_packages/npm/iconv-lite@0.4.15` to `excludes` list to continue down the rabbit hole, which brings me to the next error where it couldn't find the `isomorphic-fetch` dependency type definition, e.g:

```
jspm_packages/npm/servicestack-client@0.0.17/src/index.ts(133,27): error TS2304: Cannot find name 'Headers'.
jspm_packages/npm/servicestack-client@0.0.17/src/index.ts(154,25): error TS2304: Cannot find name 'fetch'.
```

I thought not being able to not find `fetch` was odd as I've got `"lib": ["es2015"]` which should import ES6 types like fetch. But `isomorphic-fetch` doesn't include its own types so I needed to install it with:

```
npm install @types/isomorphic-fetch --save-dev
```

Which had the strange behavior of now returning an error saying it couldn't find `servicestack-client` anymore:

```
src/app.ts(1,35): error TS2307: Cannot find module 'servicestack-client'.
```

Where running `tsc --traceResolution` showed some strange behavior, firstly it tried to look at the `npm/servicestack-client@0.0.17/*` in paths for the `isomorphic-fetch` types but started prefixing paths with `node_modules` and `node_modules/@types` to the search paths, e.g:

```
======== Resolving module 'isomorphic-fetch' from 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/index.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'isomorphic-fetch' from 'node_modules' folder.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch.ts' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch.tsx' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch.d.ts' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch/package.json' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch/index.ts' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch/index.tsx' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/isomorphic-fetch/index.d.ts' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/@types/isomorphic-fetch.ts' does not exist.
File 'C:/src/wip/typescript/jspm_packages/npm/servicestack-client@0.0.17/src/node_modules/@types/isomorphic-fetch.tsx' does not exist.
...
```

Which it eventually found:

```
File 'C:/src/wip/typescript/node_modules/@types/isomorphic-fetch/index.d.ts' exist - use it as a name resolution result.
Resolving real path for 'C:/src/wip/typescript/node_modules/@types/isomorphic-fetch/index.d.ts', result 'C:/src/wip/typescript/node_modules/@types/isomorphic-fetch/index.d.ts'
======== Module name 'isomorphic-fetch' was successfully resolved to 'C:/src/wip/typescript/node_modules/@types/isomorphic-fetch/index.d.ts'. ========
```

But it didn't bother looking at the search paths at all for resolving the `servicestack-client` module, i.e:

```
======== Resolving module 'servicestack-client' from 'C:/src/wip/typescript/src/app.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'servicestack-client' from 'node_modules' folder.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client.tsx' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client.d.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client/package.json' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client/index.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client/index.tsx' does not exist.
File 'C:/src/wip/typescript/src/node_modules/servicestack-client/index.d.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client.tsx' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client.d.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client/package.json' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client/index.ts' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client/index.tsx' does not exist.
File 'C:/src/wip/typescript/src/node_modules/@types/servicestack-client/index.d.ts' does not exist.
...
```

Which as a result ends with the error:

```
======== Module name 'servicestack-client' was not resolved. ========
```

Basically blocking this solution path at this point.

Anyway, these built-in generic TypeScript features aren't just going to work for JSPM, there's going to need to be explicit Q/A and first-class support for it to work with JSPM, basically where tsc is able to understand jspm's config.js and be able to resolve modules that way. Either that or TypeScript team could develop/maintain their own module loader so we can finally have an integrated/working solution for publishing, installing, referencing and bundling TypeScript modules - which in my experience is by far the weakest/fragile area of developing with TypeScript. 

In the meantime I'll continue down the `devDependencies` path which looks like the cleanest workaround with JSPM.
